gatsby/content/blog/2021/12/2021-12-10-sydent-typing-2.mdx

```diff
@@ -0,0 +1,387 @@
+
+Within a given module, I'd try to fix easier errors first: partly for confidence, partly to build up momentum, and partly to get myself familiar with that piece of source code. For example, I'd often start by telling mypy that `mylist = []` was actually was a `List[str]` (rather than the generic `List[Any]` which it would use otherwise).
+
+Picking and choosing easy targets works fairly well, but sometimes that means you end fixing an error that's really a symptom of an earlier one. Other times fixing one error, e.g. by giving a return type annotation to a function—would solve a series of other errors throughout the file. Watching the total number of errors mypy reports bob up and down was intriguing!
+
+In retrospect, I think it would be smoother to generate some kind of dependency graph for the package. I'm imagining a [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) where whose vertices are modules, and there's an edge `A -> B` if `A` imports from `B`. The leaves of this DAG are the ideal place to start: you can get something typechecked there without having to annotate a long chain of dependencies. Another strategy would be to see which modules were the least precise according to mypy's reports—but more on those [next time](2021-12-03-sydent-typing-1.mdx).
+
+### Annotations when working with twisted
+
+Twisted is a big project, with a few quirks that make it trickier to annotate applications using it.
+
+#### Partially typed modules and stubs
+
+Early into the process, mypy reported that calling the function [`twisted.python.log.err`](https://twistedmatrix.com/documents/current/api/twisted.python.log.html#err) was an error. It did so because I was running mypy in [`--strict` mode](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-strict). We'll talk more about why I did so and what this means [next time](2021-12-17-sydent-typing-3.mdx); for now, it's enough to know that calling a function that isn't fully annotated from within a function that is constitutes an error. `twisted` is partially annotated: many key modules and functions have type annotations, but others don't. I was reluctant to give up on `--strict`. Instead, I decided to stub the `err` function myself. 
+
+I [added](https://github.com/matrix-org/sydent/pull/418/commits/c4e04256c150ecc771e6c2049bcf52c7aaffbbe0) a `stubs` directory so that I could provide an annotation for . My initial motivation was that `mypy --strict` considers calling an untyped function from a typed function to be an error. But this was almost an accident: mypy can be configured with a list of modules to ignore if it can't find type annotations for them. `twisted` wasn't part of that list! (We'll talk more about `--strict` in the [final part](2021-12-17-sydent-typing-3.mdx) of this series.)
```

Comment:
Missing a word after "an annotation"?

Reply:
Good spot. I think the entire paragraph on line 41 is redundant, actually; meant to be replaced with that on line 39.